p4-16/spec/P4-16-draft-spec.mdk

```diff
@@ -4628,25 +4628,24 @@ the standard error ```error.ParserTimeout```.
 # Control blocks { #sec-control }
 
 P4 parsers are responsible for extracting bits from a packet into
-headers. The headers can be manipulated and transformed within ```control```
-blocks. Control blocks are a P4 construct that are used
-for describing match-action pipelines. The body of a control block
-resembles a traditional C program. Within the body of a control block,
+headers. These headers (and other metadata) can be manipulated and transformed within ```control```
+blocks. The body of a control block
+resembles a traditional imperative program. Within the body of a control block,
 match-action units can be invoked to perform data
 transformations. Match-action units are represented in P4 by
 constructs called ```tables```.
 
-There is no exceptional control-flow in a ```control``` block: no
-equivalent of the ```verify``` parser statement or of the ```reject```
-state. Error handling has to be performed explicitly by users.
-
-P4 forbids the instantiation of ```parser``` instances within a ```control```
-block.
-
-A ```control``` block may start with declarations of ```action```s, ```table```s,
-constants, variables, and instantiations.
+Syntactically, a ```control``` block is declared with a name,
+parameters, optional type parameters, and a sequence of declarations
+of constants, variables, ```action```s, ```table```s, and other
+instantiations:
 
 ~ Begin P4Grammar
+controlTypeDeclaration
```

Comment:
controlTypeDeclaration has its own section, it does not need to be duplicated here.